File: Programmers/[150369] 택배 배달과 수거하기/ONE/150369.java
class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        long sumOfDistance = 0L;
        int deliver = 0;
        int pickup = 0;
        for (int index = n - 1; index >= 0; index--) {
            if (deliveries[index] == 0 && pickups[index] == 0) {
                continue;
            }
            int count = 0;
            while (deliver < deliveries[index] || pickup < pickups[index]) {
                ++count;
                deliver += cap;
                pickup += cap;
            }
            deliver -= deliveries[index];
            pickup -= pickups[index];
            sumOfDistance += (index + 1) * 2L * count;
        }
        return sumOfDistance;
    }
}

File: Programmers/[150369] 택배 배달과 수거하기/KingDonggyu/150369.js
function initCourierBox(boxes) {
  while (!!boxes.length && !boxes[boxes.length - 1]) {
    boxes.pop();
  }
}

function handleCourierBox(cap, boxes) {
  while (!!cap && !!boxes.length) {
    const index = boxes.length - 1;

    if (!!boxes[index]) {
      boxes[index] -= 1;
      cap -= 1;
    }

    if (!boxes[index]) {
      boxes.pop();
    }
  }
}

function solution(cap, n, deliveries, pickups) {
  let answer = 0;

  while (!!deliveries.length || !!pickups.length) {
    initCourierBox(deliveries);
    initCourierBox(pickups);

    const maxDistance = Math.max(deliveries.length, pickups.length);

    handleCourierBox(cap, deliveries);
    handleCourierBox(cap, pickups);

    answer += maxDistance * 2;
  }

  return answer;
}
